src/ui/__tests__/ui_app.js

```diff
@@ -0,0 +1,263 @@
+
+        expect('linkToLastSubUrl' in newApp).to.be(true);
+        expect(isUndefined(newApp.linkToLastSubUrl)).to.be(true);
+      });
+
+      it('default navLink', () => {
+        expect(newApp.navLink).to.eql({
+          id: 'uiapp-test-defaults',
+          title: undefined,
+          order: 0,
+          url: '/app/uiapp-test-defaults',
+          description: undefined,
+          icon: undefined,
+          linkToLastSubUrl: true,
+          hidden: false,
+          disabled: false,
+          tooltip: ''
+        });
+      });
+
+      it('default order of 0', () => {
+        expect(newApp.order).to.be(0);
+      });
+
+      it('template name is ui_app', () => {
+        expect(newApp.templateName).to.be('ui_app');
+      });
+    });
+
+    describe('with spec', () => {
+      const spec = getMockSpec();
+      let newApp;
+      beforeEach(() => {
+        newApp = new UiApp(uiExports, spec);
+      });
+
+      it('use the spec ID', () => {
+        expect(newApp.id).to.be('uiapp-test');
+      });
+
+      it('copies field values from spec', () => {
+        // test that the fields exist, but have undefined value
+        expect(newApp.main).to.be('main.js');
+        expect(newApp.title).to.be('UIApp Test');
+        expect(newApp.description).to.be('Test of UI App Constructor');
+        expect(newApp.icon).to.be('ui_app_test.svg');
+        expect(newApp.linkToLastSubUrl).to.be(true);
+        expect(newApp.templateName).to.be('ui_app_test');
+        expect(newApp.order).to.be(9000);
+        expect(newApp.navLink).to.eql({
+          id: 'uiapp-test',
+          title: 'UIApp Test',
+          order: 9000,
+          url: '/app/uiapp-test',
+          description: 'Test of UI App Constructor',
+          icon: 'ui_app_test.svg',
+          linkToLastSubUrl: true,
+          hidden: false,
+          disabled: false,
+          tooltip: ''
+        });
+      });
+    });
+
+    describe('reference fields', () => {
+      const spec = getMockSpec({ testSpec: true });
+      let newApp;
+      beforeEach(() => {
+        newApp = new UiApp(uiExports, spec);
+      });
+
+      it('uiExports', () => {
+        expect(newApp.uiExports).to.be(uiExports);
+      });
+      it('spec', () => {
+        expect(newApp.spec).to.be(spec);
+      });
+    });
+
+    describe('getInjectedVars', () => {
+      it('noop function by default', () => {
+        const spec = {
+          id: 'uiapp-test'
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.getInjectedVars).to.be(noop);
+      });
+      it('reference to spec.injectVars', () => {
+        const helloFunction = () => 'hello';
+        const spec = {
+          id: 'uiapp-test',
+          injectVars: helloFunction
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.getInjectedVars).to.be(helloFunction);
+      });
+    });
+
+    describe('hidden and listed', () => {
```

Comment:
When I was reading these tests, I was having a hard time figuring out how these two properties interacted. After I read the source, I saw that listed was dependent on hidden, and that hidden was cast to a boolean values.

I think this structure makes that relationship a little clearer. What do you think?

```javascript
    describe('hidden', () => {
      describe('is cast to boolean value', () => {
        it('when undefined', () => {
          const spec = {
            id: 'uiapp-test',
          };
          const newApp = new UiApp(uiExports, spec);
          expect(newApp.hidden).to.be(false);
        });

        it('when null', () => {
          const spec = {
            id: 'uiapp-test',
            hidden: null,
          };
          const newApp = new UiApp(uiExports, spec);
          expect(newApp.hidden).to.be(false);
        });
      });
    });

    describe('listed', () => {
      it('when null defaults to the opposite value of hidden, when hidden is true', () => {
        const spec = {
          id: 'uiapp-test',
          hidden: true,
          listed: null,
        };
        const newApp = new UiApp(uiExports, spec);
        expect(newApp.listed).to.be(false);
      });

      it('when null defaults to the opposite value of hidden, when hidden is false', () => {
        const spec = {
          id: 'uiapp-test',
          hidden: false,
          listed: null,
        };
        const newApp = new UiApp(uiExports, spec);
        expect(newApp.listed).to.be(true);
      });

      it('when undefined defaults to the opposite value of hidden, when hidden is false', () => {
        const spec = {
          id: 'uiapp-test',
          hidden: false,
        };
        const newApp = new UiApp(uiExports, spec);
        expect(newApp.listed).to.be(true);
      });

      it('when undefined defaults to the opposite value of hidden, when hidden is true', () => {
        const spec = {
          id: 'uiapp-test',
          hidden: true,
        };
        const newApp = new UiApp(uiExports, spec);
        expect(newApp.listed).to.be(false);
      });

      it('when true is set to true', () => {
        const spec = {
          id: 'uiapp-test',
          listed: true,
        };
        const newApp = new UiApp(uiExports, spec);
        expect(newApp.listed).to.be(true);
      });

      it('when false is set to false', () => {
        const spec = {
          id: 'uiapp-test',
          listed: false,
        };
        const newApp = new UiApp(uiExports, spec);
        expect(newApp.listed).to.be(false);
      });
    });
```